middleware/grpc_instrumentation.go

```diff
@@ -70,47 +71,77 @@ func StreamClientInstrumentInterceptor(metric *prometheus.HistogramVec) grpc.Str
 	) (grpc.ClientStream, error) {
 		start := time.Now()
 		stream, err := streamer(ctx, desc, cc, method, opts...)
-		return &instrumentedClientStream{
-			metric:       metric,
-			start:        start,
-			method:       method,
-			ClientStream: stream,
-		}, err
+		s := &instrumentedClientStream{
+			metric:        metric,
+			start:         start,
+			method:        method,
+			serverStreams: desc.ServerStreams,
+			finished:      atomic.NewBool(false),
+			finishedChan:  make(chan struct{}),
+			ClientStream:  stream,
+		}
+		s.awaitCompletion(ctx)
+		return s, err
 	}
 }
 
+// This implementation is heavily inspired by github.com/opentracing-contrib/go-grpc's openTracingClientStream.
 type instrumentedClientStream struct {
-	metric *prometheus.HistogramVec
-	start  time.Time
-	method string
+	metric        *prometheus.HistogramVec
+	start         time.Time
+	method        string
+	serverStreams bool
+	finished      *atomic.Bool
+	finishedChan  chan struct{}
 	grpc.ClientStream
 }
 
+func (s *instrumentedClientStream) awaitCompletion(ctx context.Context) {
+	go func() {
+		select {
+		case <-s.finishedChan:
+			// Stream has finished for another reason, nothing more to do.
+		case <-ctx.Done():
+			s.finish(ctx.Err())
+		}
+	}()
+}
+
+func (s *instrumentedClientStream) finish(err error) {
+	if !s.finished.CompareAndSwap(false, true) {
+		return
+	}
+
+	close(s.finishedChan)
+
+	s.metric.WithLabelValues(s.method, errorCode(err)).Observe(time.Since(s.start).Seconds())
+}
+
 func (s *instrumentedClientStream) SendMsg(m interface{}) error {
 	err := s.ClientStream.SendMsg(m)
 	if err == nil {
 		return nil
 	}
 
-	if err == io.EOF {
-		s.metric.WithLabelValues(s.method, errorCode(nil)).Observe(time.Since(s.start).Seconds())
-	} else {
-		s.metric.WithLabelValues(s.method, errorCode(err)).Observe(time.Since(s.start).Seconds())
-	}
-
+	s.finish(err)
```

Comment:
I think this was a misunderstanding from our side. The does for `SendMsg` say:

> On error, SendMsg aborts the stream. If the error was generated by the client, the status is returned directly; otherwise, io.EOF is returned and the status of the stream may be discovered using RecvMsg.

So I think any error returned by `SendMsg` means the stream has failed. 

One improvement I can see on re-reading the docs is that if `SendMsg` returns `io.EOF`, we shouldn't call `s.finish()` yet, as the true error will be returned by `RecvMsg`, and might have a different error code than `io.EOF`. I'll make this change.

Reply:
Changed in dbb177d